Finance Nuget Shared repo

FaisalHackney opened the pull request:
For user story HFS-617

This is the initiation of the source code repo for the /finance-shared repo which will be the source code for the nuget package.

FaisalHackney committed:
First commit

Added namespaces, classes and some tests

FaisalHackney committed:
Docker file

Added text to Readme

FaisalHackney committed:
addition of the docker-yaml

added missing docker-compose yaml file

Duslerke commented on Hackney.Shared.Finance.Tests/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+COPY /nuget.config /root/.nuget/NuGet/NuGet.Config
+
+RUN dotnet restore ./Hackney.Shared.Finance/Hackney.Shared.Finance.csproj
+RUN dotnet restore ./Hackney.Shared.Finance.Tests/Hackney.Shared.Finance.Tests.csproj
```
The `dotnet build` on Line 22 already does an implicit `dotnet restore`, so we don't need line 16 an 17.

FaisalHackney replied:
removed lines 16-17

Duslerke commented on Hackney.Shared.Finance.Tests/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+COPY ./Hackney.Shared.Finance.sln ./
+COPY ./Hackney.Shared.Finance/Hackney.Shared.Finance.csproj ./Hackney.Shared.Finance/
+COPY ./Hackney.Shared.Finance.Tests/Hackney.Shared.Finance.Tests.csproj ./Hackney.Shared.Finance.Tests/
+COPY /nuget.config /root/.nuget/NuGet/NuGet.Config
```
The `COPY . . ` already copies everything specified on lines 10-14. So we don't need these explicit copy statements.

FaisalHackney replied:
Removed these COPY statements

Duslerke commented on Hackney.Shared.Finance.Tests/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+ENV DOTNET_CLI_TELEMETRY_OPTOUT='true'
+
+ARG LBHPACKAGESTOKEN
+ENV LBHPACKAGESTOKEN=$LBHPACKAGESTOKEN
```
I'm pretty sure the Line 7 is redundant.
As can be seen from this line: `LBHPACKAGESTOKEN=$LBHPACKAGESTOKEN`, the LBHPACKAGESTOKEN is already available as an environment variable as it's being used with the $ variable access syntax.

FaisalHackney replied:
removed line 7

Duslerke commented on Hackney.Shared.Finance.Tests/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+﻿FROM mcr.microsoft.com/dotnet/core/sdk:3.1
```
The base docker image does not match our project's dotnet version.
Interestingly enough, this doesn't seem to be an issue for the build process. :/

FaisalHackney replied:
Changed to `mcr.microsoft.com/dotnet/core/sdk:6.0`

Duslerke commented on Hackney.Shared.Finance.Tests/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Copy everything else and build
+COPY . .
+
+RUN dotnet build -c debug -o out Hackney.Shared.Finance.Tests/Hackney.Shared.Finance.Tests.csproj
```
The `-c` flag seems to be `debug` by default.

FaisalHackney replied:
Changed to 
`RUN dotnet build -o out Hackney.Shared.Finance.Tests/Hackney.Shared.Finance.Tests.csproj`

Duslerke commented on Hackney.Shared.Finance/Boundary/Request/PagingParametersRequest.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Hackney.Shared.Finance.Boundary.Request
+{
+    public abstract class PagingParametersRequest
```
I know I was the one who specified this to be an `abstract` class, 
but I no longer think it was the correct thing to do.

This class being `abstract` prevents it from being used as a standalone request object. There can easily be a case, where we have a 'Get Multiple Xs` endpoint that does not  have any filter parameters, but which would benefit from paging.

If we either believe that it's possible that we'll run into such case, or we want to share this with the wider team, then it may be better to increase the flexibility of this package, by making this class a regular class.

FaisalHackney replied:
Removed abstract for the use case you've mentioned

Duslerke commented on Hackney.Shared.Finance/Boundary/Response/ApiResponse.cs:
```diff
@@ -0,0 +1,25 @@
+﻿
+namespace Hackney.Shared.Finance.Boundary.Response
+{
+    /// <summary>
+    /// API Response wrapper for all API responses
+    /// If a request has been successful this will be denoted by the statusCode
+    ///     Then the 'data' property will be populated
+    /// If a request has not been successful denoted
+    ///     Then the Error property will be populated
```
The summary doesn't match the model.
Though, it's probably for the better. It may be best to stick to the standard responses dotnet itself gives for the errors. I don't think there's any value in having 1 bloated object that can contain both the error and the data.

FaisalHackney replied:
Removed the spurious comments

Duslerke commented on Hackney.Shared.Finance/Boundary/Response/ApiResponse.cs:
```diff
@@ -0,0 +1,25 @@
+    /// If a request has not been successful denoted
+    ///     Then the Error property will be populated
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ApiResponse<T> where T : class
+    {
+        public T Results { get; set; }
+
+        public long Total { get; set; }
+
+        public ApiResponse() { }
+
+        public ApiResponse(T result)
+        {
+            Results = result;
+        }
```
Actually never mind the comment above, this seems like this may be some default code from some Nuget template. Or am I wrong?

FaisalHackney replied:
No it was copied over from the Transacrtion-API codebase

Duslerke commented on Hackney.Shared.Finance/Boundary/Response/PagingParametersDomain.cs:
```diff
@@ -0,0 +1,9 @@
+﻿
+namespace Hackney.Shared.Finance.Boundary.Response
+{
+    public abstract class PagingParametersDomain
```
Same as mentioned previously. I believe these classes shouldn't be abstract classes.

FaisalHackney replied:
Removed all abstract keyword from Class declarations

Duslerke commented on Hackney.Shared.Finance/Boundary/Response/ResponseWrapper.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            get
+            {
+                if (this.Value is IEnumerable<object>)
+                {
+                    return !(this.Value as IEnumerable<object>).Any();
+                }
```
Nice code, though, is there any case where we'd want to do such check before returning an IActionResult response?

From what I remember from REST, upon querying an existing collection _(say Assets)_ with filter parameters, `even if no results are found`, the standard expected endpoint response is `200 Ok` with an empty collection: `[ ]`. Which represents the fact that collection exists, but it's either empty, or it contains nothing matching your filter parameters.

The only time the `GET multiple things` endpoint is expected to give a `404 Not Found` back is when the collection itself doesn't exist - so we don't even have the collection name specified to even query against.

So, for the above reasons, I believe that there is no circumstance, where we'd want to this **Line 14-17** check.

FaisalHackney replied:
This code does have a use case in the Transactions-api.

In the pattern we have used, IEnumerable<T> collections are returned from the Gateway (Repository) layer for those methods that are expected to return a collection of data. An `IEnumerble.Any()` query is still checking on an IEnumerable. 

If you have a method which returns an `IEnumerable<T>` and the internal code of that method has found no records, then it is far more desirable to return an `IEnumable.Empty<T>` than a null. 

An empty collection is far more representative of a **404 Not Found** than null.

Duslerke commented on Hackney.Shared.Finance/Infrastructure/Specifications/Specification.cs:
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq.Expressions;
+
+namespace Hackney.Shared.Finance.Infrastructure.Specifications
+{
+    public abstract class Specification<T> : ISpecification<T> where T : class
+    {
+        protected Specification([NotNull] Expression<Func<T, bool>> criteria)
+        {
+            Criteria = criteria;
+        }
+
+        public Expression<Func<T, bool>> Criteria { get; }
+    }
+}
```
I'm not familiar with this object. Is this something that comes with Nuget template, or is this used anywhere?

FaisalHackney replied:
It is taken from Transactions-Api and is a way to declare a predicate for an Expression<T> for querying

Duslerke commented on Hackney.Shared.Finance.Tests/Boundary/Response/ResponseWrapperTests.cs:
```diff
@@ -0,0 +1,54 @@
+
+
+        [Fact]
+        public void Given_A_Response_Entity_WhenNull_Then_ResposeWrapper_ShouldBeEmpty()
```
Method naming convention on the tests within this file.
I believe dotnet throws exceptions for: "you shouldn't use underscores within method names". At least, it does so within the API and listener projects.
Was this error suppressed, or do we not have 'Treat warnings as errors" enabled?

FaisalHackney replied:
I didn't realise that the no underscore rule in method names has been imposed on API and Listener projects. It's certainly not suppressed natively. I have removed the underscores

Duslerke commented on Hackney.Shared.Finance/Boundary/Response/ResponseWrapper.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Hackney.Shared.Finance.Boundary.Response
+{
+    public class ResponseWrapper<T> where T : class
```
The name of wrapper doesn't reflect well the fact that there **may be** no result.

Obviously, I'm pretty biased towards the `Maybe<T>` as I thought it made method signatures pretty transparent with the fact that they may return no data.

If there's hesitation about the `Maybe` name, perhaps, we can come up with an **alternative name** that represents the duality of the response state?

Duslerke replied:
Like `MaybeEmpty<T>` or `PerhapsEmpty<T>` or `PossiblyEmpty<T>`?
Perhaps, you have more ideas?

FaisalHackney replied:
Personally I think it's fine. "Maybe" and "Perhaps" match the usecase now but in the future, we could be extending this class to check for other conditions like, for example, IsChild or something. In which case, keeping the name generic seems to be the correct way to go

Duslerke reviewed (commented):
The Dockerfile is obviously an extreme nitpick - it clearly works as is.
I've just marked the places that can be trimmed down if wanted.

Also this PR contains a couple of objects I haven't seen myself. Their names aren't self-explanatory. So I'm unsure of whether we're using them, or whether it's something that's part of the Nuget template.

Lastly, there are a couple of changes I believe are pretty important:
1. Paging classes shouldn't be abstract to increase flexibility
2. We shouldn't be checking for IEnumerable being empty as there's no use case for it
3. The responseWrapper name isn't self-descriptive.

FaisalHackney committed:
PR Comments )Liudvikas)

changes appiied from PR

FaisalHackney committed:
Refactoring

Remove the IEnumerable check from the ResponseWrapper
Move Pagination classes to their own top-level namespace
